x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java

```diff
@@ -0,0 +1,550 @@
+            BytesRef upper = upperTerm == null ? null : indexedValueForSearch(upperTerm);
+            byte[] lowerBytes = lower == null ? MIN_VALUE : Arrays.copyOfRange(lower.bytes, lower.offset, 16);
+            byte[] upperBytes = upper == null ? MAX_VALUE : Arrays.copyOfRange(upper.bytes, upper.offset, 16);
+
+            // point query on the 16byte prefix
+            Query pointPrefixQuery = BinaryPoint.newRangeQuery(name(), lowerBytes, upperBytes);
+
+            Query termQuery = new TermRangeQuery(
+                name(),
+                lowerTerm == null ? null : lower,
+                upperTerm == null ? null : upper,
+                includeLower,
+                includeUpper
+            );
+
+            return new BooleanQuery.Builder().add(new BooleanClause(pointPrefixQuery, Occur.MUST))
+                .add(new BooleanClause(termQuery, Occur.MUST))
+                .build();
+        }
+    }
+
+    private boolean storeMalformed;
+    private String nullValue;
+    private BooleanFieldMapper prereleaseSubField;
+    private NumberFieldMapper majorVersionSubField;
+    private NumberFieldMapper minorVersionSubField;
+    private NumberFieldMapper patchVersionSubField;
+
+    private VersionStringFieldMapper(
+        String simpleName,
+        FieldType fieldType,
+        MappedFieldType mappedFieldType,
+        boolean storeMalformed,
+        String nullValue,
+        MultiFields multiFields,
+        CopyTo copyTo,
+        BooleanFieldMapper preReleaseMapper,
+        NumberFieldMapper majorVersionMapper,
+        NumberFieldMapper minorVersionMapper,
+        NumberFieldMapper patchVersionMapper
+    ) {
+        super(simpleName, fieldType, mappedFieldType, multiFields, copyTo);
+        this.storeMalformed = storeMalformed;
+        this.nullValue = nullValue;
+        this.prereleaseSubField = preReleaseMapper;
+        this.majorVersionSubField = majorVersionMapper;
+        this.minorVersionSubField = minorVersionMapper;
+        this.patchVersionSubField = patchVersionMapper;
+    }
+
+    @Override
+    public VersionStringFieldType fieldType() {
+        return (VersionStringFieldType) super.fieldType();
+    }
+
+    @Override
+    protected String contentType() {
+        return CONTENT_TYPE;
+    }
+
+    @Override
+    protected VersionStringFieldMapper clone() {
+        return (VersionStringFieldMapper) super.clone();
+    }
+
+    @Override
+    protected void parseCreateField(ParseContext context) throws IOException {
+        String versionString;
+        if (context.externalValueSet()) {
+            versionString = context.externalValue().toString();
+        } else {
+            XContentParser parser = context.parser();
+            if (parser.currentToken() == XContentParser.Token.VALUE_NULL) {
+                versionString = nullValue;
+            } else {
+                versionString = parser.textOrNull();
```

Comment:
Do we want to assign `nullValue` here as as well in cases where a string is empty or null?

Reply:
I agree that default values are probably not that meaningful here, but e.g. `0.0.0` might be something users might want to index in case this value is null. I think I added this for versions so they behave as close to "keyword" fields as possible, also on the indexing side, but I'm open to discuss and remove this if you think its not required.

Reply:
It would be great to discuss (maybe through a different channel?), I've been wondering in general about whether most field types should support `null_value`.

Reply:
I changed this PR to not allow setting `null_value`, however we accept `null` as a field value in the input json in which case we don't index any value into the field.